File: week-2/assignment-2/part-1/section-2/main.js
function multiplierFactory(factor) {
  let count = 0; // enclosed variable

  return {
    multiplier: function (value) {
      count++;
      return value * factor;
    },
    getCount: function () {
      return count;
    },
  };
}

function analyzeType(data) {
  if (typeof data === "number") {
    return `Number: ${data}`;
  } else if (typeof data === "string") {
    return `String: ${data}`;
  } else {
    return "Unknown Type";
  }
}

const factory = multiplierFactory(3);
console.log(factory.multiplier(5)); // 15
console.log(factory.multiplier(7)); // 21
console.log(`Function called ${factory.getCount()} times`);

const items = [10, "hello", 3.14, [1, 2]];
items.forEach((item) => {
  console.log(`Input: ${item}, Analyzed: ${analyzeType(item)}`);
});

File: week-2/assignment-2/part-1/section-2/main.py
def multiplier_factory(factor):
    count = 0  # enclosed variable in closure
    
    def multiplier(value):
        nonlocal count
        count += 1
        return value * factor
    
    def get_count():
        return count
    
    return multiplier, get_count

def analyze_type(data):
    if isinstance(data, (int, float)):
        return f"Number: {data}"
    elif isinstance(data, str):
        return f"String: {data}"
    else:
        return "Unknown Type"

if __name__ == "__main__":
    multiply_by_3, get_calls = multiplier_factory(3)
    print(multiply_by_3(5))         # 15
    print(multiply_by_3(7))         # 21
    print(f"Function called {get_calls()} times")  # 2
    
    # Dynamic typing demonstration
    for item in [10, "hello", 3.14, [1, 2]]:
        print(f"Input: {item}, Analyzed: {analyze_type(item)}")
